lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp

```diff
@@ -801,6 +799,103 @@ static void convertReturnOp(Operation *oldOp, FModuleOp topModuleOp,
   rewriter.eraseOp(oldOp);
 }
 
+//===----------------------------------------------------------------------===//
+// Handshake and StdExpr Builder classes
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+class HandshakeBuilder : public HandshakeVisitor<HandshakeBuilder, bool> {
+public:
+  HandshakeBuilder(ValueVectorList portList, Location insertLoc,
+                   ConversionPatternRewriter &rewriter)
+      : portList(portList), insertLoc(insertLoc), rewriter(rewriter) {}
+
+  bool visitInvalidOp(Operation *op) { return false; }
+  using HandshakeVisitor::visitHandshake;
+
+#define HANDLE(OPTYPE)                                                         \
+  bool visitHandshake(OPTYPE op) {                                             \
+    return buildOpLogic(portList, insertLoc, rewriter, &op), true;             \
+  }
+
+  HANDLE(handshake::BranchOp);
+  HANDLE(BufferOp);
+  HANDLE(ConditionalBranchOp);
+  HANDLE(handshake::ConstantOp);
+  HANDLE(ControlMergeOp);
+  HANDLE(ForkOp);
+  HANDLE(JoinOp);
+  HANDLE(LazyForkOp);
+  HANDLE(MergeOp);
+  HANDLE(MuxOp);
+  HANDLE(SinkOp);
+#undef HANDLE
+
+private:
+  ValueVectorList portList;
+  Location insertLoc;
+  ConversionPatternRewriter &rewriter;
+}; // namespace
+
+class StdExprBuilder : public StdExprVisitor<StdExprBuilder, bool> {
+public:
+  StdExprBuilder(ValueVectorList portList, Location insertLoc,
+                 ConversionPatternRewriter &rewriter)
+      : portList(portList), insertLoc(insertLoc), rewriter(rewriter) {}
+
+  bool visitInvalidOp(Operation *op) { return false; }
+  using StdExprVisitor::visitStdExpr;
+
+#define HANDLE(OPTYPE, FIRRTLTYPE)                                             \
+  bool visitStdExpr(OPTYPE op) {                                               \
+    return buildBinaryLogic<FIRRTLTYPE>(portList, insertLoc, rewriter), true;  \
+  }
+
```

Comment:
Personally, I'd rather do this with a template method and a template trait for mapping the argument type to the transformed type.

Reply:
Is it something like this? I feel the current approach already looks pretty clean though.
```
template <typename S>
struct FIRRTLTypeMap;

template <>
struct FIRRTLTypeMap<AddIOp> {
  typedef AddPrimOp T;
};
template <>
struct FIRRTLTypeMap<SubIOp> {
  typedef SubPrimOp T;
};
```

Reply:
I tend to avoid the preprocessor if there's another good way to do it.  But yes, something like above.

Reply:
I don't feel particularly strongly about this however, it's just my personal preference.

Reply:
Got it. How about we keep to use the current approach for this PR?